add tuple support

sahara-ooga opened the pull request:
Support tuple.

```swift
let pretty = Pretty(formatter: SinglelineFormatter())
let tuple = (1, ("one", URL(string: "https://www.example.com/")!))
print(pretty.string(tuple, debug: false))    
//(1, ("one", https://www.example.com/))

/* label */
let labeledTuple = (2019, region: "Chili", variety: Optional("Chardonnay"), taste: ["round", "smooth", "young"])
print(pretty.string(labeledTuple, debug: false))
//(2019, region: "Chili", variety: "Chardonnay", taste: ["round", "smooth", "young"])
```

this solves #34.

sahara-ooga committed:
add tuple support

sahara-ooga committed:
add formatter tests

sahara-ooga committed:
refactoring

sahara-ooga committed:
Merge branch 'master' into feature/tuple

# Conflicts:
#	Sources/Core/Formatter/MultilineFormatter.swift

YusukeHosonuma commented on Sources/Core/Pretty.swift:
```diff
@@ -22,7 +22,7 @@ struct Pretty {
         let mirror = Mirror(reflecting: target)
 
-        // Optional / Collection / Dictionary
+        // Optional / Collection / Dictionary / Tuple
```
I deleted this line in #70 because understand from `switch` statements.
Please delete for prevent conflict 🙏

sahara-ooga replied:
I delete the comment in c9b082f👍🏻

YusukeHosonuma commented on Tests/Core/Formatter/SinglelineFormatterTests.swift:
```diff
@@ -31,7 +32,15 @@ class SinglelineFormatterTests: XCTestCase {
+    
+    func testTupleString() {
+        assert(to: formatter.tupleString).expect([
+            when([], then: #"()"#),
```
Good 👍

YusukeHosonuma commented on Tests/Core/Formatter/SinglelineFormatterTests.swift:
```diff
@@ -31,7 +32,15 @@ class SinglelineFormatterTests: XCTestCase {
+        assert(to: formatter.tupleString).expect([
+            when([], then: #"()"#),
+            when(["\"one\""], then: #"("one")"#),
+            when(["\"one\"", "\"two\""], then: #"("one", "two")"#),
```
I suggest to include 2 types in tuple (e.g. `["\"one\"", "2"]`) that like representation of tuple

sahara-ooga replied:
That`s a good idea!

sahara-ooga replied:
@YusukeHosonuma I followed your suggestion in 6d7c912👍🏻

YusukeHosonuma commented on Tests/Core/PrettyTests.swift:
```diff
@@ -132,6 +132,19 @@ class PrettyTests: XCTestCase {
+    /// Tuple
+    ///
+    func testString_Tuple() {
+        let tuple = (1, ("one", URL(string: "https://www.example.com/")!))
```
Labeled tuple (e.g. `(one: 1, two: "two")`) is not support in this PR?

sahara-ooga replied:
I'll try it.

sahara-ooga replied:
Label is supported in d43f963.

YusukeHosonuma reviewed (commented):
Good works!!

sahara-ooga commented on Sources/Core/Pretty.swift:
```diff
@@ -22,7 +22,7 @@ struct Pretty {
         let mirror = Mirror(reflecting: target)
 
-        // Optional / Collection / Dictionary
+        // Optional / Collection / Dictionary / Tuple
```
```suggestion
```

sahara-ooga committed:
remove a comment.

Apply suggestions from code review.

sahara-ooga committed:
image poli-type handling...

thanks to Yusuke, I get the image that we can show the merit to support tuple type. A tuple can hold multi type element, which is distinct from Array.

sahara-ooga committed:
show tuple label

YusukeHosonuma commented on Tests/Core/Formatter/MultilineFormatterTests.swift:
```diff
@@ -66,6 +66,35 @@ class MultilineFormatterTests: XCTestCase {
+            One(value: 1,
+                first: true)
+            """)
+        ]
```
What about a this code style?

```suggestion
        let tupleElements: [(String?, String)] = [
            (
                ".0",
                "\"first\""
            ),
            (
                "label",
                """
                One(value: 1,
                    first: true)
                """
            )
        ]
```

sahara-ooga replied:
I adopt it in b10f02b, thx

YusukeHosonuma commented on Tests/Core/Formatter/MultilineFormatterTests.swift:
```diff
@@ -66,6 +66,35 @@ class MultilineFormatterTests: XCTestCase {
     }
+    
+    func testTupleString() {
+        let tupleElements: [(String?, String)] = [
```
What time first argument of tuple is `nil`?
(Please write to test that passed to `nil` if possible)

sahara-ooga replied:
I can not find a situation in which the label is `nil`.

# Example

## the tuple is `()`

a formatter receives `[]`.

## no label, one element

the tuple is transformed to the inner element instance.

`(#""one""#)` becomes `"one"`.

## no label element and labeled element

`(#""one""#, second: 2)` becomes `[(".0", "\"\"one\"\""), ("second", "2")]`.

YusukeHosonuma commented on Tests/Core/PrettyTests.swift:
```diff
@@ -132,6 +132,27 @@ class PrettyTests: XCTestCase {
+                when(true, then: #"(1, ("one", URL("https://www.example.com/")))"#),
+            ])
+        
+        let labeledTuple = (vintage: 2019, region: "Chili", variety: Optional("Chardonnay"), taste: ["round", "smooth", "young"])
```
I think a better test would be include unlabeled data 🍷 

e.g.  `(2019, "Chili", variety: Optional("Chardonnay"), taste: ["round", "smooth", "young"])`

sahara-ooga replied:
I adopt it in daf0d9d3892eb690843267b1465e9c9927c92018.🍇

YusukeHosonuma commented on Sources/Core/Formatter/SinglelineFormatter.swift:
```diff
@@ -15,6 +15,20 @@ class SinglelineFormatter: PrettyFormatter {
+        // if the labels of tuples are not specificated, they are like ".1" (not nil).
+        // Using "." as the first charactor of the label of tuple is prohibited.
+        let labelValuePairs: [String] = elements.map { label, value in
+            if let label = label, label.first != "." {
```
I think a better way would be this judge move to `Pretty.swift`.

For example:

```swift
// Pretty.swift
case .tuple:
    let elements: [(String?, String)] = mirror.children.map {
        var labelOrNil: String? = nil
        if let label = $0.label, label.first != "." {
            labelOrNil = label
        }
        return (label: labelOrNil, value: _string($0.value))
    }
    return formatter.tupleString(elements: elements)

// Here
let labelValuePairs: [String] = elements.map { label, value in
    if let label = label {
...
```

Formatter should be concentrate styling therefore should not judge property of data.

sahara-ooga replied:
Thank you for a comment!

This suggestion is so reasonable, I adopt it in b4676ced4298929c3309c5711ace37317e273de6.

YusukeHosonuma commented on Sources/Core/Formatter/SinglelineFormatter.swift:
```diff
@@ -15,6 +15,20 @@ class SinglelineFormatter: PrettyFormatter {
+            }
+        }
+
+        return "(\(labelValuePairs.joined(separator: ", ")))"
```
Is this easier to read?

```suggestion
        return "(" + labelValuePairs.joined(separator: ", ") + ")"
```

YusukeHosonuma reviewed (commented):
Good works!!

sahara-ooga committed:
correct tuple style

followed @YusukeHosonuma's suggestion

Co-Authored-By: Yusuke Hosonuma <tobi462@gmail.com>

sahara-ooga committed:
add non label element for test example

sahara-ooga committed:
move the judgement of label from formatter to pretty

along with it, modify test cases of single line formatter because formatters receive proper labels.

sahara-ooga committed:
refactoring: modify string consolidation

Co-Authored-By: Yusuke Hosonuma <tobi462@gmail.com>

sahara-ooga committed:
Merge branch 'master' into feature/tuple

# Conflicts:
#	Sources/Core/Pretty.swift
#	Tests/Core/PrettyTests.swift

YusukeHosonuma commented on Sources/Core/Formatter/MultilineFormatter.swift:
```diff
@@ -32,6 +32,24 @@ class MultilineFormatter: PrettyFormatter {
+        // if the labels of tuples are not specificated, they are like ".1" (not nil).
+        // Using "." as the first charactor of the label of tuple is prohibited.
+        let labelValuePairs: [String] = elements.map { label, value in
+            if let label = label, label.first != "." {
```
Missed in b4676ce maybe :)

```suggestion
        let labelValuePairs: [String] = elements.map { label, value in
            if let label = label {
```

sahara-ooga replied:
Thank you! I applied it. b9f5a83 8990689

YusukeHosonuma reviewed (approved):
Thank you for your polite works ❤️ 
Please check last suggestion comment from me :)

That's all!!

sahara-ooga committed:
fix to remove leaked unnecessary judge

Co-Authored-By: Yusuke Hosonuma <tobi462@gmail.com>

sahara-ooga committed:
fix a test case

sahara-ooga commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Tuple対応](https://trello.com/c/9JEA1Axp/1237-tuple%E5%AF%BE%E5%BF%9C)
